detox/android/detox/src/full/java/com/wix/detox/espresso/hierarchy/ViewHierarchyGenerator.kt

```diff
@@ -0,0 +1,128 @@
+package com.wix.detox.espresso.hierarchy
+
+import android.util.Xml
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import com.wix.detox.reactnative.ui.getAccessibilityLabel
+import org.xmlpull.v1.XmlSerializer
+import java.io.StringWriter
+
+object ViewHierarchyGenerator {
+    @JvmStatic
+    fun generateXml(shouldInjectTestIds: Boolean): String {
+        val rootViews = RootViewsHelper.getRootViews()
+        return generateXmlFromViews(rootViews, shouldInjectTestIds)
+    }
+
+    private fun generateXmlFromViews(rootViews: List<View?>?, shouldInjectTestIds: Boolean): String {
+        return StringWriter().use { writer ->
+            val serializer = Xml.newSerializer().apply {
+                setOutput(writer)
+                startDocument(Xml.Encoding.UTF_8.name, true)
+                setFeature("http://xmlpull.org/v1/doc/features.html#indent-output", true)
+                startTag("", "ViewHierarchy")
+            }
+
+            rootViews?.forEach { rootView ->
+                rootView?.let {
+                    serializeViewHierarchy(it, serializer, shouldInjectTestIds, emptyList())
+                }
+            }
+
+            serializer.apply {
+                endTag("", "ViewHierarchy")
+                endDocument()
+            }
+
+            writer.toString()
```

Comment:
good point, it does has `flush()` method since it inherits from `Write` (`flush` is `abstract`), but it means nothing this case since it immediately writes to an in-memory resource (the string), so there's [nothing to flush](https://github.com/openjdk/jdk11/blob/master/src/java.base/share/classes/java/io/StringWriter.java#L230-L234)